2023/Q3/artifacts/PoS-quint/helper-functions.qnt

```diff
@@ -845,153 +872,133 @@ module helperFunctions {
                             finalRate: int,
-                            slashedAmountsMap: Epoch -> int): Epoch -> int = {
+                            slashesMap: Epoch -> int): Epoch -> int = {
       val infractionEpoch = curEpoch - slashProcessingDelay
-      val initTotalUnbonded = (infractionEpoch+1).to(curEpoch).fold(0, (sum, e) => sum + computeTotalUnbonded(validator, infractionEpoch, slashes, totalUnbonded.get(e), totalRedelegatedUnbonded.get(e)))
-      val initBalanceBonds = (infractionEpoch+1).to(curEpoch).fold(0, (sum, e) => sum + (totalBonded.get(e) - computeRecentTotalUnbonded(infractionEpoch, totalUnbonded.get(e), totalRedelegatedUnbonded.get(e))))
-      val zeroedBalanceRedelegatedBonds = (curEpoch+1).to(curEpoch+PIPELINE_OFFSET).mapBy(e => 0)
-      val initBalanceRedelegatedBonds = (infractionEpoch+1).to(curEpoch).fold(zeroedBalanceRedelegatedBonds, (acc, e) => computeBalanceRedelegatedBonds(totalRedelegatedBonded.get(e), e, curEpoch, slashes, totalRedelegatedUnbonded, acc))
-      // We must iterate in epoch increasing order.
-      val epochRange = range(curEpoch+1, curEpoch+PIPELINE_OFFSET+1)
-      val result = epochRange.foldl((initTotalUnbonded,
-                                     initBalanceBonds,
-                                     initBalanceRedelegatedBonds,
-                                     slashedAmountsMap), (acc, e) => val updatedTotalUnbonded = acc._1 + computeTotalUnbonded(validator, infractionEpoch, slashes, totalUnbonded.get(e), totalRedelegatedUnbonded.get(e))
-                                                                     val updatedBalanceBonds = acc._2 + (totalBonded.get(e) - computeRecentTotalUnbonded(infractionEpoch, totalUnbonded.get(e), totalRedelegatedUnbonded.get(e)))
-                                                                     val updatedBalanceRedelegatedBonds = computeBalanceRedelegatedBonds(totalRedelegatedBonded.get(e), e, curEpoch, slashes, totalRedelegatedUnbonded, acc._3)
-                                                                     val slashedAmount = (infractionStake - updatedTotalUnbonded) * finalRate
-                                                                     val currentStake = stakes.get(e) - acc._4.get(e)
-                                                                     val slashableStake = (currentStake - updatedBalanceBonds - updatedBalanceRedelegatedBonds.get(e))
-                                                                     (updatedTotalUnbonded, updatedBalanceBonds, updatedBalanceRedelegatedBonds, acc._4.set(e, acc._4.get(e) + min(slashedAmount, slashableStake))))
-      result._4
+      val bonds = totalBonded.keys().filter(e => e <= infractionEpoch).mapBy(start => totalBonded.get(start))
+      val redelegatedBonds = bonds.keys().mapBy(e => totalRedelegatedBonded.getOrElse(e, Map()))
+      val epochRange = (curEpoch+1).to(curEpoch+PIPELINE_OFFSET).sortSetDecreasing()
```

Comment:
previously this ranged from `curEpoch+1` to `curEpoch+PIPELINE_OFFSET+1`, why is it only to `curEpoch+PIPELINE_OFFSET` now?

Reply:
Before it was also ranging from `curEpoch+PIPELINE_OFFSET` as the interval generated by `range(curEpoch+1, curEpoch+PIPELINE_OFFSET+1)` is open on the high endpoint `curEpoch+PIPELINE_OFFSET+1`. See [Quint's syntax cheetsheet](https://github.com/informalsystems/quint/blob/main/doc/quint-cheatsheet.pdf).